File: Information_Search_and_Textmining/[과제 B1] 한국어 형태소 분석기 개발/과제B1_한국어형태소분석기개발_인공지능학부20243153백여민.py
#----------------------
# 정보검색과 텍스트마이닝 :: [과제 B1] 한국어 형태소 분석기 개발
# 인공지능학부 20243153 백여민
#----------------------
'''
josaEomi.py에서 !가 제거 되지 않아,
"훔쳤다!" 및 "적이다!" 등이 아래와 같이 분리되지 않는 문제 해결했습니다. 

훔쳤다
        훔쳤 + 다/조사
        훔쳤 + 다/어미
 적이다
        적 + 이다/조사
        적이 + 다/조사
        적이 + 다/어미       
        
또한, 조사 / 어미 분리가 되지 않는 경우,
==> 조사/어미 분리되지 않음 이라고 출력이 되도록 했습니다. 
'''

import re
from pathlib import Path

BASE = Path(__file__).resolve().parent

def load_list(relpath: str):
    path = BASE / relpath
    items = []
    with open(path, 'r', encoding='cp949') as f:
        for line in f:
            s = line.strip()
            if not s:
                continue
            items.append(s)
    # 긴 항목부터 우선 매칭하도록 길이 기준 내림차순
    items.sort(key=len, reverse=True)
    return items

def remove_punc(s: str) -> str:
    # 최소한의 문장부호 제거(원 코드와 유사)
    punc = ['\n','\r','\'','"','-','=','/','.', '(', ')', ',', '!', '?', ':', ';']
    for ch in punc:
        s = s.replace(ch, '')
    return s

def find_all_suffixes(word: str, suffixes, tag: str):
    """
    word가 suffix로 끝나면 stem + suffix를 모두 찾아서 출력.
    return값은 발견 여부 (bool)
    """
    found = False
    for suf in suffixes:
        if len(suf) >= len(word):
            continue
        if word.endswith(suf):
            stem = word[:-len(suf)]
            if stem:  # stem이 비어있지 않을 때만
                print(f"\t{stem} + {suf}/{tag}")
                found = True
    return found

def analyze_line(line: str, josa, eomi):
    line = remove_punc(line)
    words = [w for w in line.split() if w]
    for w in words:
        print(w)
        # 조사 먼저 시도 → 실패 시 어미
        j_found = find_all_suffixes(w, josa, "조사")
        e_found = find_all_suffixes(w, eomi, "어미")
        if not (j_found or e_found):
            print("\t==> 조사/어미 분리되지 않음")

def main():
    josa = load_list("josa96.txt")
    eomi = load_list("eomi152.txt")
    test_path = BASE / "test.txt"

    with open(test_path, 'r', encoding='cp949') as f:
        for line in f:
            if len(line.strip()) < 1:
                continue
            analyze_line(line, josa, eomi)

if __name__ == "__main__":
    main()

File: Information_Search_and_Textmining/[과제 B1] 한국어 형태소 분석기 개발/과제B1_한국어형태소분석기개발_인공지능학부20243153백여민.cpp
//실행 전 동일한 파일 경로에 필요한 텍스트파일들이 존재하는지 확인하세요!

#include <bits/stdc++.h>
using namespace std;

vector<string> load_list(const string& filename) {
    vector<string> items;
    ifstream fin(filename);
    if (!fin.is_open()) {
        cerr << "파일 열기 실패: " << filename << "\n";
        return items;
    }
    string line;
    while (getline(fin, line)) {
        // 끝 공백 제거
        while (!line.empty() && (line.back() == '\r' || line.back() == '\n'))
            line.pop_back();
        if (!line.empty()) items.push_back(line);
    }
    fin.close();
    // 긴 항목부터 매칭되게 정렬
    sort(items.begin(), items.end(), [](const string& a, const string& b) {
        return a.size() > b.size();
        });
    return items;
}

string remove_punc(const string& s) {
    static const string punc = "\n\r'\"-=/.,()!?:;";
    string out = s;
    for (char c : punc) {
        out.erase(remove(out.begin(), out.end(), c), out.end());
    }
    return out;
}

bool find_all_suffixes(const string& word, const vector<string>& suffixes, const string& tag) {
    bool found = false;
    for (const auto& suf : suffixes) {
        if (suf.size() >= word.size()) continue;
        if (word.size() >= suf.size() &&
            equal(suf.rbegin(), suf.rend(), word.rbegin())) { // ends_with
            string stem = word.substr(0, word.size() - suf.size());
            if (!stem.empty()) {
                cout << "\t" << stem << " + " << suf << "/" << tag << "\n";
                found = true;
            }
        }
    }
    return found;
}

void analyze_line(const string& line, const vector<string>& josa, const vector<string>& eomi) {
    string cleaned = remove_punc(line);
    istringstream iss(cleaned);
    string word;
    while (iss >> word) {
        cout << word << "\n";
        bool j_found = find_all_suffixes(word, josa, "조사");
        bool e_found = find_all_suffixes(word, eomi, "어미");
        if (!j_found && !e_found) {
            cout << "\t==> 조사/어미 분리되지 않음\n";
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    // 실행 경로에 txt 파일이 있다고 가정
    vector<string> josa = load_list("josa96.txt");
    vector<string> eomi = load_list("eomi152.txt");

    ifstream fin("test.txt");
    if (!fin.is_open()) {
        cerr << "test.txt 파일을 열 수 없습니다.\n";
        return 1;
    }

    string line;
    while (getline(fin, line)) {
        if (line.find_first_not_of(" \t\r\n") == string::npos) continue; // 빈 줄 스킵
        analyze_line(line, josa, eomi);
    }
    fin.close();
    return 0;
}
